### DISCUSSION TITLE
Is it possible to control the tracing::instrument(ret) level, to make it different from the span's?

### DISCUSSION THREAD

**Discussion by scullionw** [tokio-rs/tracing#2328] [Q&A] [answered]
```rust
#[tracing::instrument(ret, level = "debug")]
pub fn myfn(context: String) -> String {
    warn!("careful!");

    todo!()
}
```

i would like the return value to be logged on the "DEBUG" level only, to cut back on the logging noise. However, I want the warn!(...) log here to have the instrumented context attached to it. Right now, the context is only attached when the level is debug.

**Comment by hawkw** [comment, marked as answer]
This isn't currently possible, but it would be a good feature request! We could add something like
```rust
#[tracing::instrument(ret(level = "debug"), level = "warn")]
fn whatever() -> ... {
   ...
}
```
to allow changing the level for the `ret` event.

We would probably want to do it by adding a `ret(level = "<level>")` rather than just `ret = "<level>"` to future-proof the macro interface, so that we can add additional configuration arguments beyond setting the return event's level.

I think this should be a fairly straightforward change, at least for someone who's familiar with procedural macros, if you're interested in implementing it?

**Reply by davidbarsky** [reply]
@hawkw I think I answered something similar, but I think I threw out a bikeshed a month or two back: https://github.com/tokio-rs/tracing/discussions/2290#discussioncomment-3510585. Do you have any thoughts on that approach?

**Comment by shekhirin** [comment]
For future generations: yes, it is possible since 0.1.24 https://github.com/tokio-rs/tracing/blob/c297a37096ae1562d453624984aa5f924ab490a1/tracing-attributes/CHANGELOG.md?plain=1#L92-L103
